clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -10129,6 +10129,8 @@ def note_within_field_of_type : Note<
   "within field of type %0 declared here">;
 def note_illegal_field_declared_here : Note<
   "field of illegal %select{type|pointer type}0 %1 declared here">;
+def note_illegal_type_decl_here : Note<
+  "declaration %0 of illegal type %1 is here">;
```

Comment:
```suggestion
  "declaration %0 of invalid type %1 is here">;
```
Let's not block on this if it turns out to be a contentious suggestion. We usually try to avoid using "illegal" in diagnostics because that has implications in different parts of the world, but we don't always succeed (as the preceding note shows).

Actually, I don't think we need this new note at all. We can dodge it entirely by using `note_declared_at` from `DiagnosticCommonKinds.td`.

Reply:
> We usually try to avoid using "illegal" in diagnostics because that has implications in different parts of the world

Thank you for noting this. I didn't know about that.

> Actually, I don't think we need this new note at all. 

I know it is probably not essential, but I wanted to show a concrete type and declaration's to the user in the message, I think it looks much clearer, especially when declaration of unsupported type depends on some template parameters.
If you still think that it is better to use `note_declared_at` I ll switch to it.

Reply:
We try to reuse existing diagnostics as much as possible because we can (and have, repeatedly) run out of diagnostic IDs, so my default as a reviewer is "how can we use an existing diagnostic instead of adding a new one". While I think your wording is more precise than "declared here" and looks better in isolation, notes are associated with a specific warning or error diagnostic already. So, to me, it really boils down to whether we think the %0 and %1 are necessary. Based on the test cases, I don't think they are because they all look like this:
```
type decl; // declaration 'decl' of illegal type 'type' is here
```
where the diagnostic is repeating what's already on the line of code. Given that I don't think declaring zero-sized arrays is a common thing, and that it's not likely you'll get multiple declarations on the same line, I think the existing note is sufficiently clear and we should go with it. WDYT?

(Btw, the first time somebody complains about the note being confusing, I'm fully on board with your proposed wording.)

Reply:
And, done.